### ISSUE TITLE
how many components are there in the Physics Derivation graph?

### ISSUE THREAD

**Issue by researcherben** [allofphysicsgraph/proofofconcept#241]
Is the content in the Physics Derivation Graph [connected](https://en.wikipedia.org/wiki/Component_(graph_theory))?

If there is more than one independent subgraph, then there is a connecting derivation to be found.

**Comment by bhpayne** [comment]
The technical jargon for this is that I want to count the number of Weakly Connected Components (WCC)
<https://neo4j.com/docs/graph-data-science/current/algorithms/wcc/>

BLUF: There's a query that uses the Graph Data Science (GDS) library. While a Cypher query that doesn't involve GDS can be made, Gemini doesn't recommend a pure Cypher query due to performance and complexity. I confirmed the GDS library does exist (!) -- <https://neo4j.com/docs/graph-data-science/current/>

The page <https://neo4j.com/blog/graph-algorithms-neo4j-weakly-connected-components/> provided an example that I then fed into Gemini <https://aistudio.google.com/>

_Here's my prompt_:

I'm using Neo4j to store nodes and edges of a property graph. 

The Weakly Connected Components, or Union Find, algorithm finds sets of connected nodes. I want to count the number of connected components in the graph.

The following Cypher produces a graph with 6 nodes and 2 weakly connected components:
```cypher
MERGE (nAlice:User {id:"Alice"})
MERGE (nBridget:User {id:"Bridget"})
MERGE (nCharles:User {id:"Charles"})
MERGE (nDoug:User {id:"Doug"})
MERGE (nMark:User {id:"Mark"})
MERGE (nMichael:User {id:"Michael"})
MERGE (nAlice)-[:FRIEND]->(nBridget)
MERGE (nAlice)-[:FRIEND]->(nCharles)
MERGE (nMark)-[:FRIEND]->(nDoug)
MERGE (nMark)-[:FRIEND]->(nMichael);
```
What Cypher query when run against that data would count the number of connected components? Provide the Cypher query 

<HR>

_Here's Gemini's response_:

```cypher
MATCH (n)
WITH collect(id(n)) AS ids
CALL gds.wcc.stream({
  nodeProjection: '*',
  relationshipProjection: {
    FRIEND: {
      type: 'FRIEND',
      properties: {}
    }
  },
  nodeIds: ids
})
YIELD componentId
WITH collect(componentId) AS componentIds
RETURN size(componentIds) AS numberOfComponents
```

**Explanation:**

1.  **`MATCH (n)`**: This matches all nodes in the graph.
2.  **`WITH collect(id(n)) AS ids`**: This collects the IDs of all nodes into a list called `ids`.
3.  **`CALL gds.wcc.stream(...) YIELD componentId`**:
    *   This calls the Neo4j Graph Data Science (GDS) library's Weakly Connected Components algorithm (`gds.wcc.stream`).
    *   `nodeProjection: '*'`: This means we are projecting all nodes in the graph.
    *   `relationshipProjection: { FRIEND: { type: 'FRIEND', properties: {} } }`: This specifies that we are only considering relationships of type `FRIEND` when determining connected components. We don't care about the properties of this relationship, hence the `{}`.
    *   `nodeIds: ids`:  This tells GDS to run the algorithm only on the nodes we collected in step 2.
    *   `YIELD componentId`: For each node, the algorithm returns a `componentId` that identifies which component the node belongs to.
4.  **`WITH collect(componentId) AS componentIds`**:  This collects all the `componentId` values into a list called `componentIds`. Note that nodes in the same component will have the same `componentId`.
5.  **`RETURN size(componentIds) AS numberOfComponents`**: The `size()` function returns the number of items in the list, which will be the number of distinct component IDs, effectively giving us the number of connected components.

**How It Works:**

The Weakly Connected Components algorithm identifies groups of nodes where any node can reach any other node within the group, following the direction of a relationship if a directed relationship. In the context of an undirected graph such as the one created in the initial Cypher statement, any node in a component can reach any other node in that component by following edges.  The algorithm assigns each node in the same component to the same `componentId`. Because we collect the `componentId` values and then use `size()` to get the number of unique component IDs, this gives the number of connected components.

**To use this query you need to have the Neo4j Graph Data Science (GDS) library installed in your Neo4j server.**

<HR>

_My follow-up question_:

Is there a way to accomplish the count of Weakly Connected Components in a Neo4j graph without using the Graph Data Science (GDS) library?

<HR>

_Gemini's answer_:

Yes, you can determine the count of weakly connected components without the GDS library, although it's generally less efficient for larger graphs. The approach involves using a recursive common table expression (CTE) to traverse the graph and identify all reachable nodes from a given starting node. Then, we iterate over nodes to find unvisited ones to count distinct components. Here's the Cypher query:

```cypher
WITH collect(id(n)) AS allNodeIds
WITH allNodeIds, [x IN allNodeIds | {id: x, visited: false}] AS nodeStates
WITH allNodeIds, nodeStates

UNWIND allNodeIds AS startNodeId
WITH allNodeIds, nodeStates, startNodeId
WHERE NOT nodeStates[ apoc.coll.indexOf(allNodeIds,startNodeId)].visited
WITH allNodeIds, nodeStates, startNodeId
CALL {
    WITH startNodeId, allNodeIds, nodeStates
    
    MATCH (start) WHERE id(start) = startNodeId
    
    WITH start, allNodeIds, nodeStates
    
        
    CALL {
        WITH start, allNodeIds, nodeStates
        
        MATCH path = (start)-[:FRIEND*]-(reachable)
        WHERE all(x IN nodes(path) WHERE id(x) IN allNodeIds)
        
        WITH collect(DISTINCT reachable) AS componentNodes, allNodeIds, nodeStates
        RETURN componentNodes
    }

    WITH componentNodes, startNodeId, nodeStates, allNodeIds
    UNWIND componentNodes AS node
    WITH allNodeIds, nodeStates, startNodeId, node
    WITH allNodeIds, nodeStates, startNodeId, node, apoc.coll.indexOf(allNodeIds,id(node)) AS idx
    WITH  allNodeIds, apoc.coll.set(nodeStates, idx, {id:id(node), visited: true}) AS updatedNodeStates, startNodeId
    RETURN updatedNodeStates AS nodeStates
}
WITH nodeStates
RETURN size(filter(x IN nodeStates WHERE x.visited=true)) as totalVisitedNodes,
       size(nodeStates)-size(filter(x IN nodeStates WHERE x.visited=true)) as unvisitedNodeCount,
	   size(filter(x IN nodeStates WHERE x.visited=true and x.componentStart = true )) as componentCount

```

**Explanation:**

1. **Initialization:**
   - `WITH collect(id(n)) AS allNodeIds`: Collects all node IDs into `allNodeIds`.
   - `WITH allNodeIds, [x IN allNodeIds | {id: x, visited: false}] AS nodeStates`: Creates a list of node state records, each with an ID and a `visited` flag, initialized to `false`.

2. **Iterating over unvisited nodes:**
   - `UNWIND allNodeIds AS startNodeId`: Iterates over the node IDs.
   - `WHERE NOT nodeStates[ apoc.coll.indexOf(allNodeIds,startNodeId)].visited`: Filters to only process node IDs that haven't been visited.
   -The `CALL` block encapsulates the logic for finding a component:

3. **Finding connected components:**
    - Find the start node based on `startNodeId`.
    - Uses `MATCH path = (start)-[:FRIEND*]-(reachable)` to find all reachable nodes connected by any length of `:FRIEND` relationships.  `WHERE all(x IN nodes(path) WHERE id(x) IN allNodeIds)` checks that all nodes in path are from the set of all nodes we're iterating over.
    - It collects the distinct reachable nodes of a given component in `componentNodes`

4. **Updating node states:**
  - The `componentNodes` are unwound and the `visited` field of the given node's record in `nodeStates` is set to true.

5.  **Returning Results:**
   - It returns the size of nodes marked as `visited` as `totalVisitedNodes`.  It also returns `unvisitedNodeCount` and `componentCount`.  Note that `componentCount` is also equal to the number of components.

**How it works:**

*   It loops through each node in the graph and if the node has not been visited then the inner `CALL` will traverse all connected nodes by following the `:FRIEND` relationship and updating a boolean value indicating the node has been visited.
*   The algorithm iterates through all the nodes of the graph and each time it encounters an unvisited node, it starts a new traversal and increments the component counter by 1. This ensures that all nodes that are reachable from the node via `:FRIEND` edges are marked as visited within a given component.

**Important Considerations:**

*   **Performance:** This approach, especially the recursive part, is significantly less efficient than the GDS implementation, especially on large graphs because it does not have a parallel or optimized implementation. The GDS algorithm is designed specifically for graph algorithms and is highly optimized.
*   **Complexity:** This approach has higher time complexity compared to GDS, which can be crucial when handling large graphs. The performance will likely degrade exponentially with increasing graph size.
*   **Readability:** This implementation is more verbose and less readable compared to the GDS call.

**When to use it:**

*   You might consider this approach if you absolutely cannot use the GDS library (e.g., due to licensing or environment constraints)
*   This method would be suitable only for small graphs where performance is not critical.
*   It could also serve as an educational tool to understand how the WCC algorithm works.

**In Summary:**

While it's possible to achieve the same outcome as the GDS weakly connected components algorithm without it using a recursive CTE, it's generally advisable to use the GDS library due to its performance advantages for real-world use cases, unless you have very specific limitations. The GDS solution is simpler, more concise, and much faster.
